packages/local_auth/lib/local_auth.dart

```diff
@@ -90,6 +90,19 @@ class LocalAuthentication {
         'authenticateWithBiometrics', args);
   }
 
+  /// Returns true if auth was cancelled successfully.
+  /// Returns false if there was some error or no auth in progress.
+  ///
+  /// Returns [Future] bool true or false:
+  Future<bool> stopAuthentication() {
```

Comment:
Out of curiosity, how would the Dart app decide to call this?

Reply:
For devices with in-display fingerprint sensor, when authentication is in progress and if user clicks on any other button on the screen the fingerprint scanning will still be in progress. Hence, we needed an API to stop the authentication in such scenarios.

Otherwise, the fingerprint scanner stops working and the user has to restart the application.

Reply:
For devices with in-display fingerprint sensor(assuming that authentication is in progress), if user clicks on any other button on the screen, the fingerprint scanning will still be in progress. Hence, we needed an API to stop the authentication in such scenarios.

Reply:
Thanks for the response!

> If user clicks on any other button on the screen

This is the part that I don't get. Why would I allow user to click anything in my app if there's authentication in progress? The user is *not* authenticated so should not be interacting with the app at all. I was assuming that on-screen auth will show exactly the same dialog. The only difference would be where you touch.

Could you please point me to a video or an article of some sort which shows cancelAuthentication in action?

Reply:
@mehmetf For in-display fingerprint scanners, we don't get any popup. We just get the fingerprint scanner icon.

http://tiny.cc/e345ez

*Assuming current behaviour of the plugin*
In the link, you can see that we only get the fingerprint scanner icon and no popup. The user can still type the passcode, and click done. When user clicks done, and we navigate to another screen the authentication is still going on. The plugin doesn't provide any mechanism to cancel this authentication.

I hope I was clear in explaining. :)

Reply:
Very interesting.

Something is broken in this flow though.

- The image shows FP and a PIN and user can choose to authenticate either way.
- If both of these are *not* appearing due to this plugin then who is displaying the PIN entry method? That seems wrong. So, I assume both of these are appearing due to the plugin calling `biometric.authenticate`.
- So, assuming biometric API is showing the PIN entry, if user enters PIN instead of FP, why would `biometric.authenticate` not return "success"? Why would we still expect a fingerprint scan?

I am OK adding this API because it is supported on Android. However the use case you are describing seems very very wrong to me. The app should not be responsible for cancelling authentication *if the user already authenticated*.

If you agree with me, could you take some time to research how the Android API is supposed to be used? In particular why would PIN entry not run success/failure callbacks? (https://github.com/flutter/plugins/blob/master/packages/local_auth/android/src/main/java/io/flutter/plugins/localauth/AuthenticationHelper.java#L142)

Reply:
@mehmetf Thanks for your quick response.

1. The screenshot is not from my app but explains the exact scenario we need in our app. The screenshot is from oneplus's app locker screen.
2. The PIN entry is not due to the plugin. It's the app's screen having print entry, and ```biometric.authenticate``` just overlays the fingerprint icon on top of it for scanning (The user can still interact with other part of the screen though).
3. Because the biometric API is not responsible for showing the PIN entry, the biometric.authenticate doesn't know anything about it. There's no relation between them. :)

I'm not sure if the use case is wrong, but I have seen this flow in a lot of apps.

Reply:
I see!

OK that makes more sense. So basically you app is offering two auth methods simultaneously and you want to cancel one if the user chooses the other one.

Reply:
@mehmetf Exactly. :)